uom/common/container/kdtree.h

```diff
@@ -0,0 +1,627 @@
+struct KdTree<PointType>::TreeNode : Node {
+  // The left child of *this.
+  NodeUniquePtr left = nullptr;
+  // The right child of *this.
+  NodeUniquePtr right = nullptr;
+  // The parent of *this, not owned.
+  TreeNode* parent = nullptr;
+
+  // Gether structure information from left and right sub-trees.
+  void PushUp();
+  // Apply structure information to left and right sub-trees.
+  void PushDown();
+  // Returns true if the node hasn't childs.
+  bool IsLeaf() const { return left == nullptr && right == nullptr; }
+  // Returns the distance from node range to given point.
+  double RangeTo(const PointType& query_point) const;
+};
+
+template <typename PointType>
+struct KdTree<PointType>::NearestNeighborSearcher {
+  NearestNeighborSearcher(const PointType& query_point, int k) : k(k), query_point(query_point) {}
+
+  void Search(const NodeUniquePtr& node);
+
+  // A heap to maintain the nearest `k` neighbors.
+  using QueueItem = std::pair<const TreeNode*, double>;
+  struct QueueItemCompare {
+    bool operator()(const QueueItem& lhs, const QueueItem& rhs) const {
+      return lhs.second < rhs.second;
+    }
+  };
+  std::priority_queue<QueueItem, std::vector<QueueItem>, QueueItemCompare> queue;
+
+  // Passed parameters.
+  int k = 0;
+  const PointType& query_point;
+};
+
+template <typename PointType>
+void KdTree<PointType>::TreeNode::PushUp() {
+  this->size = 1;
+  this->range = RangeType(this->point);
+  this->num_invalid_nodes = static_cast<int>(this->point_deleted);
+
+  auto collect_from_child = [this](TreeNode* child) {
+    child->parent = this;
+    this->size += child->size;
+    this->range.extend(child->range);
+    this->num_invalid_nodes += child->num_invalid_nodes;
+  };
+
+  if (left != nullptr) {
+    collect_from_child(left.get());
+  }
+  if (right != nullptr) {
+    collect_from_child(right.get());
+  }
+
+  this->radius = ((this->range.max() - this->range.min()) * 0.5).norm();
+  this->tree_deleted =
+      this->point_deleted && (!left || left->tree_deleted) && (!right || right->tree_deleted);
+}
+
+template <typename PointType>
+void KdTree<PointType>::TreeNode::PushDown() {
+  auto apply_to_child = [this](TreeNode* child) {
+    child->tree_deleted |= this->tree_deleted;
+    child->point_deleted |= child->tree_deleted;
+    if (this->tree_deleted) {
+      child->num_invalid_nodes = child->size;
+    }
+  };
+
+  if (left != nullptr) {
+    apply_to_child(left.get());
+  }
+  if (right != nullptr) {
+    apply_to_child(right.get());
+  }
+}
+
+template <typename PointType>
+double KdTree<PointType>::TreeNode::RangeTo(const PointType& query_point) const {
+  double squared_distance = 0;
+  for (int d = 0; d < kDim; ++d) {
+    const double val = query_point[d];
+    const double min_val = this->range.min()[d];
+    const double max_val = this->range.max()[d];
+    if (val < min_val) {
+      squared_distance += (val - min_val) * (val - min_val);
+    }
+    if (val > max_val) {
```

Comment:
if `min_val `<= `min_val`, then `val` can not be < `min_val `and > `max_val ` at the same time. add `else`?

Reply:
change to 
```c++
    if (val <= min_val) {
      squared_distance += (val - min_val) * (val - min_val);
    } else if (val > max_val) {
      squared_distance += (val - max_val) * (val - max_val);
    }
```

`max_val ` will never less or equal to `min_val`, the only two positions to update the range can guarantee it.
```c++
// position 1
this->range = RangeType(this->point);
// position 2
this->range.extend(child->range);
```

Reply:
change to 
```c++
    if (val <= min_val) {
      squared_distance += (val - min_val) * (val - min_val);
    } else if (val > max_val) {
      squared_distance += (val - max_val) * (val - max_val);
    }
```

`max_val ` will never less or equal to `min_val`, the only two positions to update the range can guarantee it.
```c++
// position 1
this->range = RangeType(this->point);
// position 2
this->range.extend(child->range);
```